File: examples/delphi/my-app/src/Database/ProjectRepository.pas
unit Database.ProjectRepository;

interface

uses
  System.SysUtils,
  Models.Project,
  Database.Connection,
  Utils.Logger;

type
  TProjectRepository = class
  private
    FConnection: TDatabaseConnection;
    FLogger: TLogger;
  public
    constructor Create;
    destructor Destroy; override;
    
    procedure Save(AProject: TProject);
    procedure Update(AProject: TProject);
    procedure Delete(AProjectId: Integer);
    function FindById(AProjectId: Integer): TProject;
    function FindByOwnerId(AOwnerId: Integer): TArray<TProject>;
    function FindAll: TArray<TProject>;
  end;

implementation

constructor TProjectRepository.Create;
begin
  inherited;
  FConnection := TDatabaseConnection.GetInstance;
  FLogger := TLogger.Create;
end;

destructor TProjectRepository.Destroy;
begin
  FLogger.Free;
  inherited;
end;

procedure TProjectRepository.Save(AProject: TProject);
begin
  FLogger.Log('Saving project: ' + AProject.GetName);
  FConnection.ExecuteQuery('INSERT INTO projects...');
end;

procedure TProjectRepository.Update(AProject: TProject);
begin
  FLogger.Log('Updating project: ' + IntToStr(AProject.GetId));
  FConnection.ExecuteQuery('UPDATE projects...');
end;

procedure TProjectRepository.Delete(AProjectId: Integer);
begin
  FLogger.Log('Deleting project: ' + IntToStr(AProjectId));
  FConnection.ExecuteQuery('DELETE FROM projects...');
end;

function TProjectRepository.FindById(AProjectId: Integer): TProject;
begin
  FLogger.Log('Finding project by ID: ' + IntToStr(AProjectId));
  Result := TProject.Create;
end;

function TProjectRepository.FindByOwnerId(AOwnerId: Integer): TArray<TProject>;
begin
  FLogger.Log('Finding projects by owner ID: ' + IntToStr(AOwnerId));
  SetLength(Result, 0);
end;

function TProjectRepository.FindAll: TArray<TProject>;
begin
  FLogger.Log('Finding all projects');
  SetLength(Result, 0);
end;

end.

File: examples/kotlin/my-app/src/main/kotlin/com/example/repositories/ProjectRepository.kt
package com.example.repositories

import com.example.models.Project
import com.example.database.Database

class ProjectRepository {
    private val db = Database
    
    fun save(project: Project) {
        db.execute("INSERT INTO projects VALUES (?, ?, ?, ?, ?, ?)",
            project.id, project.name, project.description, project.ownerId, 
            project.createdAt, project.updatedAt)
    }
    
    fun findById(id: String): Project? {
        return db.query("SELECT * FROM projects WHERE id = ?", id)
    }
    
    fun findByOwner(ownerId: String): List<Project> {
        return db.queryAll("SELECT * FROM projects WHERE owner_id = ?", ownerId)
    }
    
    fun findAll(): List<Project> {
        return db.queryAll("SELECT * FROM projects")
    }
    
    fun update(project: Project) {
        db.execute("UPDATE projects SET name = ?, description = ?, updated_at = ? WHERE id = ?",
            project.name, project.description, project.updatedAt, project.id)
    }
    
    fun delete(id: String): Boolean {
        return db.execute("DELETE FROM projects WHERE id = ?", id) > 0
    }
}
